src/lib.rs

```diff
@@ -0,0 +1,271 @@
+//! An [RFC 1952](https://datatracker.ietf.org/doc/html/rfc1952)-correct description
+//! of the gzip file format in Rust.
+//!
+//! To use the library, create a [Reader] or [Writer] with an underlying data structure
+//! implementing [Read] or [Write], respectively. Libz stream initialization and management
+//! is carried out automatically.
+
+use std::{
+    ffi::{CStr, CString},
+    io::{self, Read, Write},
+    mem::size_of,
+    ptr::null_mut,
+};
+
+use libc::c_void;
+use libz_sys::*;
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum Mode {
+    INFLATE,
+    DEFLATE,
+}
+
+/// Safety-wrapped representation of a libz stream.
+#[repr(C)]
+struct Stream {
+    mode: Mode,
+    stream: z_stream,
+}
+
+impl Drop for Stream {
+    /// Shutdown the stream and any relevant resources in use by it.
+    fn drop(&mut self) {
+        unsafe {
+            match self.mode {
+                Mode::DEFLATE => deflateEnd(self.into()),
+                Mode::INFLATE => inflateEnd(self.into()),
+            };
+        }
+    }
+}
+
+impl AsRef<z_stream> for Stream {
+    fn as_ref(&self) -> &z_stream {
+        &self.stream
+    }
+}
+
+impl Into<z_streamp> for &mut Stream {
+    fn into(self) -> z_streamp {
+        &mut self.stream
+    }
+}
+
+impl Stream {
+    fn default_stream() -> z_stream {
+        unsafe {
+            z_stream {
+                zalloc: std::mem::transmute::<
+                    *const (),
+                    unsafe extern "C" fn(*mut c_void, u32, u32) -> *mut c_void,
+                >(null_mut() as *const ()),
+                zfree: std::mem::transmute::<
+                    *const (),
+                    unsafe extern "C" fn(*mut c_void, *mut c_void),
+                >(null_mut() as *const ()),
+                opaque: null_mut(),
```

Comment:
The default initialization values for these functions is `Z_NULL` in the documentation, but this value is not provided by `libz-sys`. Right now I make a cast, but it doesn't seem keen on cooperating.